File: android/app/src/main/java/com/hisab/app/data/stock/StockRepository.kt
package com.hisab.app.data.stock

import androidx.room.withTransaction
import com.hisab.app.data.HisabDatabase
import com.hisab.app.data.LOCAL_SHOP_ID
import com.hisab.app.data.sync.StockMovementSyncPayload
import com.hisab.app.data.sync.SyncOutboxEntity
import com.hisab.app.domain.EntityId
import com.hisab.app.domain.Quantity
import com.hisab.app.domain.StockMovement
import com.hisab.app.domain.StockMovementType
import com.hisab.app.domain.correctStock
import com.hisab.app.domain.damage
import com.hisab.app.domain.generateId
import com.hisab.app.domain.restock
import kotlinx.coroutines.flow.Flow
import java.time.Clock

/**
 * Reading stock and changing it by hand (Step 41): a delivery arriving, goods
 * lost or broken, and a shelf count.
 *
 * Every read here is a `SUM` over the movement ledger, never a stored number
 * (D002, D020). That is why a change shows on screen immediately with nothing
 * to refresh: the row that was just inserted is part of the sum the screen is
 * already watching.
 *
 * Every write queues a sync event in the same database transaction as the
 * movement (D003), so a stock change can never exist with nothing waiting to
 * send it.
 *
 * Sales write their own movements as part of one sale transaction
 * (`SaleRepository`), because a movement without its sale would be a stock
 * change nobody can explain.
 */
class StockRepository(
    private val database: HisabDatabase,
    private val clock: Clock = Clock.systemUTC(),
    private val shopId: String = LOCAL_SHOP_ID,
) {
    private val movements = database.stockMovementDao()
    private val outbox = database.syncOutboxDao()

    /** Every product with the stock its movements add up to (Step 41). */
    fun observeProductsWithStock(
        query: String = "",
        includeInactive: Boolean = false,
    ): Flow<List<ProductWithStock>> = movements.observeProductsWithStock(shopId, query.trim(), includeInactive)

    suspend fun currentStock(productId: EntityId): Quantity = Quantity(movements.currentStockScaled(productId.value))

    suspend fun historyFor(productId: EntityId): List<StockMovementEntity> = movements.forProduct(productId.value)

    /**
     * Everything that moved stock other than a sale, newest first, for the
     * history screen (Step 42). A sale's own movement is left out because the
     * sale itself is already listed — showing both would be the same event
     * twice.
     */
    fun observeRecentMovements(limit: Int = HISTORY_LIMIT): Flow<List<MovementWithProduct>> =
        movements.observeRecentMovements(StockMovementType.SALE.name, limit)

    /** Goods arriving from a supplier. `note` is free text the shopkeeper may leave blank. */
    suspend fun restock(
        productId: EntityId,
        quantity: Quantity,
        note: String? = null,
    ): StockMovement =
        save(
            restock(
                productId = productId,
                quantity = quantity,
                occurredAt = clock.instant(),
                sourceReference = cleanNote(note),
            ),
        )

    /** Goods lost, broken or expired. `quantity` is how much was lost, as a positive amount. */
    suspend fun damage(
        productId: EntityId,
        quantity: Quantity,
        note: String? = null,
    ): StockMovement =
        save(
            damage(
                productId = productId,
                quantity = quantity,
                occurredAt = clock.instant(),
                sourceReference = cleanNote(note),
            ),
        )

    /**
     * A shelf count: `counted` is what is actually there, and the movement
     * records the difference from what the ledger said (`correctStock`).
     *
     * The ledger is read inside the same transaction as the write, so a sale
     * saved at the same moment cannot slip in between the reading and the
     * correction and leave the count wrong.
     */
    suspend fun count(
        productId: EntityId,
        counted: Quantity,
        note: String? = null,
    ): StockMovement =
        database.withTransaction {
            val movement =
                correctStock(
                    productId = productId,
                    countedQuantity = counted,
                    recordedQuantity = Quantity(movements.currentStockScaled(productId.value)),
                    occurredAt = clock.instant(),
                    sourceReference = cleanNote(note),
                )
            writeWithEvent(movement)
            movement
        }

    /**
     * Stores a stand-alone movement the server sent (Step 47), or — when it is
     * one this phone made — notes that the server now has it. No sync event is
     * queued: sending it back would be an echo.
     */
    suspend fun applyFromServer(movement: StockMovement) {
        database.withTransaction {
            if (movements.byId(movement.id.value) == null) {
                movements.insert(movement.toEntity())
            } else {
                movement.time.serverReceivedAt?.let { movements.markServerReceived(movement.id.value, it) }
            }
        }
    }

    private suspend fun save(movement: StockMovement): StockMovement {
        database.withTransaction { writeWithEvent(movement) }
        return movement
    }

    private suspend fun writeWithEvent(movement: StockMovement) {
        movements.insert(movement.toEntity())
        outbox.insert(
            SyncOutboxEntity(
                eventId = generateId().value,
                entityType = ENTITY_TYPE_STOCK_MOVEMENT,
                entityId = movement.id.value,
                operation = OPERATION_CREATE,
                payload = StockMovementSyncPayload.toJson(movement),
                baseRevision = null,
                clientTimestamp = movement.time.occurredAt,
            ),
        )
    }

    private fun cleanNote(note: String?): String? = note?.trim()?.takeIf { it.isNotEmpty() }

    companion object {
        const val HISTORY_LIMIT = 100
        const val ENTITY_TYPE_STOCK_MOVEMENT = "StockMovement"
        const val OPERATION_CREATE = "create"
    }
}

File: server/src/modules/inventory/stockRepository.ts
import type { PoolClient } from 'pg'
import { getPool } from '../../db/pool.js'
import { pgErrorCode, UNIQUE_VIOLATION } from '../../db/transaction.js'
import { quantity, type Quantity } from '../../domain/quantity.js'
import type { StockMovement } from '../../domain/stock.js'
import { movementFromRow } from '../sales/saleRepository.js'

type Queryable = Pick<PoolClient, 'query'>

/**
 * Stock on the server, always as a SUM over stock_movement — never a stored
 * number (D002, D020). Every function takes the session's shopId and filters
 * on it (D015).
 */

export interface ProductStock {
  readonly productId: string
  readonly name: string
  readonly unit: string
  readonly quantity: Quantity
}

/** Every live product in the shop with the stock its movements add up to; a product that never moved reads 0. */
export async function currentStock(shopId: string): Promise<ProductStock[]> {
  const { rows } = await getPool().query<{
    id: string
    name: string
    unit: string
    stock: string
  }>(
    `SELECT p.id, p.name, p.unit,
            COALESCE((
              SELECT SUM(m.quantity_delta_scaled)
              FROM stock_movement m
              WHERE m.shop_id = p.shop_id AND m.product_id = p.id
            ), 0) AS stock
     FROM product p
     WHERE p.shop_id = $1 AND p.deleted_at IS NULL
     ORDER BY lower(p.name) ASC`,
    [shopId],
  )
  return rows.map((row) => ({
    productId: row.id,
    name: row.name,
    unit: row.unit,
    quantity: quantity(Number(row.stock)),
  }))
}

export async function stockOf(
  shopId: string,
  productId: string,
  db: Queryable = getPool(),
): Promise<Quantity> {
  const { rows } = await db.query<{ stock: string }>(
    `SELECT COALESCE(SUM(quantity_delta_scaled), 0) AS stock
     FROM stock_movement WHERE shop_id = $1 AND product_id = $2`,
    [shopId, productId],
  )
  return quantity(Number(rows[0]!.stock))
}

/** One product's movements, newest first. */
export async function movementsOf(
  shopId: string,
  productId: string,
  limit: number,
): Promise<StockMovement[]> {
  const { rows } = await getPool().query(
    `SELECT * FROM stock_movement
     WHERE shop_id = $1 AND product_id = $2
     ORDER BY occurred_at DESC, id DESC
     LIMIT $3`,
    [shopId, productId, limit],
  )
  return rows.map(movementFromRow)
}

export type MovementOutcome =
  | { readonly status: 'created'; readonly movement: StockMovement }
  /** The same id was already stored for this shop — a retry, not a second movement (D004). */
  | { readonly status: 'exists'; readonly movement: StockMovement }
  /** That id already belongs to another shop's row. */
  | { readonly status: 'refused' }

async function findMovement(
  db: Queryable,
  shopId: string,
  id: string,
): Promise<StockMovement | null> {
  const { rows } = await db.query('SELECT * FROM stock_movement WHERE shop_id = $1 AND id = $2', [
    shopId,
    id,
  ])
  return rows[0] === undefined ? null : movementFromRow(rows[0])
}

/**
 * Stores a movement that stands on its own — a restock, damage, customer
 * return or shelf count. A sale's movements are written with the sale
 * (`saveSaleTransaction`), never through here.
 */
export async function saveStandaloneMovement(
  db: Queryable,
  shopId: string,
  movement: StockMovement,
): Promise<MovementOutcome> {
  try {
    const inserted = await db.query(
      `INSERT INTO stock_movement
         (id, shop_id, product_id, movement_type, quantity_delta_scaled, source_reference, occurred_at)
       VALUES ($1, $2, $3, $4, $5, $6, $7)
       ON CONFLICT (id) DO NOTHING
       RETURNING id`,
      [
        movement.id,
        shopId,
        movement.productId,
        movement.type,
        movement.quantityDelta,
        movement.sourceReference,
        movement.time.occurredAt,
      ],
    )
    const stored = await findMovement(db, shopId, movement.id)
    if (stored === null) return { status: 'refused' }
    return { status: inserted.rowCount === 0 ? 'exists' : 'created', movement: stored }
  } catch (error) {
    if (pgErrorCode(error) === UNIQUE_VIOLATION) return { status: 'refused' }
    throw error
  }
}

/**
 * Stand-alone movements changed after a cursor, oldest first, for a pull
 * (Step 47). Movements that belong to a sale are not here: they travel inside
 * the sale.
 */
export async function changedStandaloneMovements(
  shopId: string,
  afterCursor: number,
  limit: number,
): Promise<Array<{ movement: StockMovement; seq: number }>> {
  const { rows } = await getPool().query(
    `SELECT * FROM stock_movement
     WHERE shop_id = $1 AND sale_id IS NULL AND change_seq > $2
     ORDER BY change_seq ASC
     LIMIT $3`,
    [shopId, afterCursor, limit],
  )
  return rows.map((row) => ({ movement: movementFromRow(row), seq: Number(row.change_seq) }))
}
